central/notifications/datastore/internal/writer/writer_impl.go

```diff
@@ -0,0 +1,139 @@
+package writer
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/notifications/datastore/internal/store"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/protoconv"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+var (
+	_ Writer = (*writerImpl)(nil)
+
+	rootNamespaceUUID = uuid.Nil
+)
+
+type writerImpl struct {
+	mutex sync.Mutex
+
+	buffer map[string]*storage.Notification
+	store  store.Store
+}
+
+func (c *writerImpl) read(id string) (*storage.Notification, bool) {
+	notification, found := c.buffer[id]
+	return notification, found
+}
+
+func (c *writerImpl) write(notification *storage.Notification) {
+	c.buffer[notification.GetId()] = notification
+}
+
+func (c *writerImpl) Upsert(ctx context.Context, notification *storage.Notification) error {
+	enrichedNotification := enrichNotificationWithDefaults(notification)
+	id := enrichedNotification.GetId()
+
+	var baseNotification *storage.Notification
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	notificationInBuffer, found := c.read(id)
+	// If a notification already exists in the buffer it is the most recent.
+	// We use it as a base to merge with the new notification.
+	if found {
+		baseNotification = notificationInBuffer
+	} else {
+		// If no notification is in the buffer, we try to fetch a notification
+		// from the database. If found, we use it as the base for the merge.
+		notificationInDB, found, err := c.store.Get(ctx, id)
+		if err != nil {
+			return errors.Wrap(err, "failed to query for existing record")
+		}
+		if found {
+			baseNotification = notificationInDB
+		}
+	}
+
+	// Merge notifications to up the occurrence and update the time stamps.
+	if baseNotification != nil {
+		enrichedNotification = mergeNotifications(baseNotification, enrichedNotification)
+	}
+
+	c.write(enrichedNotification)
+	return nil
+}
+
+func (c *writerImpl) Flush(ctx context.Context) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	notificationChunk := make([]*storage.Notification, 0, len(c.buffer))
+	for _, notification := range c.buffer {
+		notificationChunk = append(notificationChunk, notification)
+	}
+	c.buffer = make(map[string]*storage.Notification)
```

Comment:
Hmm, I'm not sure whether it's too eager resetting the buffer here.

In case we fail to write the notifications, all previous notifications within the buffer will be lost.

In case of a client without permissions attempting to flush notifications, it'll lead to losing all notifications. Should we just keep the buffer in case of an error, but have a fixed size for it?
The fixed size may be beneficial either way: currently nothing is stopping the buffer from growing infinitely, if clients mess up using `Flush()` properly, this could cause troubles.

Reply:
Moved the reset to below the write to make sure it's successful. Also added a buffer max size, and return retryable errors if the buffer is full.

Reply:
I was wondering if we should automagically flush the buffer if it's full vs returning retryable errors with explicit flushes by the client.

Reply:
Decided to flush the buffer automagically when it's full.